### PR TITLE
Redesign transcript search with Material UI

### PR DESCRIPTION
**Pull request by mverrey**
Redesign of the home page as a search/filter system instead of a table of contents.

No one with any design sense has done any work on this branch, and as such any guidance towards that effect would be much appreciated!

### CODE REVIEW

File: app/Transcripts.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
+'use client'
+
+import { useEffect, useState } from 'react';
+import { usePathname, useRouter, useSearchParams } from 'next/navigation';
+import Link from 'next/link';
```
Comment (awong-dev) [inline app/Transcripts.tsx:5]: nit: sort the includes

File: app/Transcripts.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      setVideos([]);
+      return () => {};
+    }
+
+    let ignore = false;
+
+    const startDateString = filters.dateRange.start !== null ? formatDateForPath(filters.dateRange.start) : null;
+    const endDateString = filters.dateRange.end !== null ? formatDateForPath(filters.dateRange.end) : null;
+
+    getAllVideosForDateRange(filters.category, startDateString, endDateString)
```
Comment (awong-dev) [inline app/Transcripts.tsx:32]: I've been trying to use await in all the code unless there's a veeeery specific reason to use the chaining syntax. Move to that?

File: app/Transcripts.tsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+function getDateFromParams(params: URLSearchParams, key: string): Date | null {
+  const param: string | null = params.get(key);
+  if (param === null) {
+    return null;
+  }
+
+  const date: Date = parseDateFromPath(param);
+
+  if (isValid(date)) {
```
Comment (awong-dev) [inline app/Transcripts.tsx:138]: nit: prefer early return null on error and last line being the success when easy.

File: app/Transcripts.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    return date;
+  }
+  else {
+    return null;
+  }
+}
+
+function buildUrl(basePath: string, filterParams: TranscriptFilterSelection): string {
```
Comment (awong-dev) [inline app/Transcripts.tsx:146]: I wonder if we could use this:

https://stackoverflow.com/a/69437289

File: app/Transcripts.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+function getDateRangeFromParams(params: URLSearchParams): DateRange {
+  const start: Date | null = getDateFromParams(params, 'start');
+  const end: Date | null = getDateFromParams(params, 'end');
+
+  if (start === null && end === null) {
+    return { start: parseDateFromPath('2024-03-01'), end: null };
+  }
+  else {
```
Comment (awong-dev) [inline app/Transcripts.tsx:125]: drop ending else{} block.

File: app/Transcripts.tsx [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    return 'sps-board';
+  }
+}
+
+function getDateRangeFromParams(params: URLSearchParams): DateRange {
+  const start: Date | null = getDateFromParams(params, 'start');
+  const end: Date | null = getDateFromParams(params, 'end');
+
+  if (start === null && end === null) {
+    return { start: parseDateFromPath('2024-03-01'), end: null };
```
Comment (awong-dev) [inline app/Transcripts.tsx:123]: Pull default date into constant pelase

File: app/Transcripts.tsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  };
+}
+
+function getCategoryFromParams(params: URLSearchParams, allCategories: string[]): string {
+  const categoryParam: string | null = params.get('category');
+
+  if (categoryParam !== null && allCategories.includes(categoryParam)) {
+    return categoryParam;
+  }
+  else {
```
Comment (awong-dev) [inline app/Transcripts.tsx:113]: nit: drop last else block.

File: components/TranscriptFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import { MenuItem, Select } from '@mui/material';
+import { DatePicker } from "@mui/x-date-pickers";
+import { isAfter, isBefore } from "date-fns";
+
+export type DateRange = {
+  start: Date | null,
```
Comment (awong-dev) [inline components/TranscriptFilter.tsx:6]: What about using 0 and something way off in the future for unset?  Then the type is simpler..

**Review by awong-dev — COMMENTED** [review]
Looking pretty good!  I like it!

Couple of comments....mostly style nits.

File: app/Transcripts.tsx (hunk 2 above)
Comment (mverrey) [reply app/Transcripts.tsx:32]: You can't await in a client component

File: app/Transcripts.tsx (hunk 4 above)
Comment (mverrey) [reply app/Transcripts.tsx:146]: Wow that looks way easier. I was trying to find any kind of URL utility functions and gave up and built a string.

File: components/TranscriptFilter.tsx (hunk 8 above)
Comment (mverrey) [reply components/TranscriptFilter.tsx:6]: I would tend to disagree. The type would be hidden, not simpler. If I can safely assume that any non-null date object is valid, then it prevents having to check at display/conversion time.

File: app/Transcripts.tsx (hunk 2 above)
Comment (mverrey) [reply app/Transcripts.tsx:32]: Here's some context for that: https://react.dev/reference/react/useEffect#fetching-data-with-effects
Basically, a client component function can't be async, and you can't pass an async function into useEffect. You can build and then call an async function from within the useEffect callback, but it's a lot of boilerplate for not a lot of value IMO.

One thing they recommend that I do think I want to do is move the fetch into a hook of its own, which seems a lot tidier.

File: app/Transcripts.tsx (hunk 6 above)
Comment (mverrey) [reply app/Transcripts.tsx:123]: Done, but the constant default is a hack IMO anyway. I want to think of a better way of coming up with a default start date. One thought is checking when the most recent transcript is from, and then subtracting a month from that.

File: app/Transcripts.tsx (hunk 4 above)
Comment (mverrey) [reply app/Transcripts.tsx:146]: Never mind, looks like that only works with the Pages router. For some reason they removed UrlObject from the App router. I can't seem to find any discussion online.

File: app/Transcripts.tsx (hunk 2 above)
Comment (mverrey) [reply app/Transcripts.tsx]: I did end up using the temporary async function when I moved the fetch into a hook, but I don't think it's any cleaner.

**Pull request merged**